### PR TITLE
MGDOBR-1125: New Instance creation ignores Error Handler definition

### PR DESCRIPTION
**Pull request by kelvah**
JIRA: https://issues.redhat.com/browse/MGDOBR-1125

The issue was only present when creating a bridge with an EH configuration without encountering validation errors.

### CODE REVIEW

**Comment by manstis** [comment]
@kelvah Strangely I get a "fatal: Need to specify how to reconcile divergent branches." error from `git` when fetching locally.

Previously, you rebasing the PR on `main` fixed it.. Could you please try that here too?

**Comment by kelvah** [comment]
> @kelvah Strangely I get a "fatal: Need to specify how to reconcile divergent branches." error from `git` when fetching locally.
> 
> Previously, you rebasing the PR on `main` fixed it.. Could you please try that here too?

@manstis sure thing. done!

File: src/app/Instance/CreateInstance/machines/createInstanceMachine.ts [hunk 1]
```diff
@@ -216,7 +217,6 @@ const createInstanceMachine = createMachine(
                   errorHandlerChange: {
                     actions: "setErrorHandler",
                     target: ".validate",
-                    cond: "isSubmitted",
                   },
```
[the remark this answers is not in the document]
Comment (manstis) [reply src/app/Instance/CreateInstance/machines/createInstanceMachine.ts:220]: Now we have the code to look at I'd love another talk about how this works.

When you did your first talk it was all a bit hypothetical, but now I _see_ it and _run_ it.. I'd love to know more.

@vpellegrino WDYT?

**Review by manstis — APPROVED** [review]
It seems to work fine.. :+1: 

I'd love to understand the XState configuration a whole lot more.

File: src/app/Instance/CreateInstance/machines/createInstanceMachine.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (vpellegrino) [reply src/app/Instance/CreateInstance/machines/createInstanceMachine.ts:220]: @manstis this is a good example, makes sense

**Review by vpellegrino — APPROVED** [review]
LGTM, thanks @kelvah

File: src/app/Instance/CreateInstance/machines/createInstanceMachine.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (kelvah) [reply src/app/Instance/CreateInstance/machines/createInstanceMachine.ts:220]: @manstis sure, I'll book a slot.

File: src/app/Instance/CreateInstance/CreateInstance.test.tsx [hunk 2]
```diff
@@ -347,6 +379,20 @@ const setName = (comp: RenderResult, name: string): void => {
   });
 };
 
+const setErrorHandlginMethod = async (
```
Comment (Srihari1192) [inline src/app/Instance/CreateInstance/CreateInstance.test.tsx:382]: const setErrorHandlginMethod = async (
```suggestion
const setErrorHandlingMethod = async (
```

File: src/app/Instance/CreateInstance/CreateInstance.test.tsx [hunk 3]
```diff
@@ -170,14 +163,53 @@ describe("CreateInstance component", () => {
     const instanceName = "my instance";
     setName(comp, instanceName);
 
-    const selector = comp.baseElement.querySelector(
-      "[data-ouia-component-id='error-handling-method-selector']"
-    );
-    fireEvent.click(
-      selector?.querySelector(".pf-c-select__toggle-arrow") as Node
+    await waitFor(() => setErrorHandlginMethod(comp, "Webhook"));
```
[the remark this answers is not in the document]
Comment (Srihari1192) [reply src/app/Instance/CreateInstance/CreateInstance.test.tsx:166]: @vpellegrino this suggested change need to be update in other places too in CreateInstance.test.tsx

**Comment by vpellegrino** [comment]
Sorry @kelvah my fault.
I accepted and committed a proposed partial change, by mistake.
I just reverted it: https://github.com/5733d9e2be6485d52ffa08870cabdee0/sandbox-ui/pull/214/commits/1ec525f8f7602b364791bc5870506bda51989e46

**Comment by kelvah** [comment]
> Sorry @kelvah my fault. I accepted and committed a proposed partial change, by mistake. I just reverted it: [1ec525f](https://github.com/5733d9e2be6485d52ffa08870cabdee0/sandbox-ui/commit/1ec525f8f7602b364791bc5870506bda51989e46)

No worries. Thanks :+1: 

I'm converting this to draft as I spotted some issues and I'm afraid I'll need to change the `ErrorHandling` component further. Let's give priority to #204.

File: src/app/Instance/CreateInstance/CreateInstance.test.tsx (hunk 3 above)
[the remark this answers is not in the document]
Comment (kelvah) [reply src/app/Instance/CreateInstance/CreateInstance.test.tsx]: Fixed now

**Review by manstis — COMMENTED** [review]
A couple of minor comments.

However I also noticed if I first try to create an Instance with an existing name (click save and get the name already taken error); correct it and then set the Error Handler to something (other than Ignore) the Error Handler form shows an error (e.g. Webhook URL not set) before I click save again. I believe this behaviour is incorrect and that the Error Handler form errors should not be shown until I click save.

**Comment by kelvah** [comment]
> A couple of minor comments.

@manstis I don't see other comments beside the one here below. :eyes: 

> 
> However I also noticed if I first try to create an Instance with an existing name (click save and get the name already taken error); correct it and then set the Error Handler to something (other than Ignore) the Error Handler form shows an error (e.g. Webhook URL not set) before I click save again. I believe this behaviour is incorrect and that the Error Handler form errors should not be shown until I click save.

Great catch! Fixed now.

I am now moving to the `unsubmitted` state when there are saving errors. After the server side validation happens, we have to "reset" the client-side one.

File: src/app/Instance/CreateInstance/machines/createInstanceMachine.ts [hunk 4]
```diff
@@ -269,32 +282,42 @@ const createInstanceMachine = createMachine(
       }),
       setErrorHandler: assign((context, { method, parameters }) => {
```
Comment (manstis) [inline src/app/Instance/CreateInstance/machines/createInstanceMachine.ts:283]: Any reason not to rename this `_context` too (for consistency)?

File: src/app/Instance/CreateInstance/machines/createInstanceMachine.ts [hunk 5]
```diff
@@ -269,32 +282,42 @@ const createInstanceMachine = createMachine(
...
-          ...context,
           errorHandler: {
+            ...context.errorHandler,
             method,
             parameters,
           },
         };
       }),
-      setCreationError: assign((context, { error }) => {
+      setErrorHandlerValidator: assign((context, { validator }) => {
```
Comment (manstis) [inline src/app/Instance/CreateInstance/machines/createInstanceMachine.ts:292]: Any reason not to rename this `_context` too (for consistency)?

File: src/app/Instance/CreateInstance/machines/createInstanceMachine.ts [hunk 6]
```diff
@@ -251,16 +266,14 @@ const createInstanceMachine = createMachine(
     actions: {
       setName: assign((context, { name }) => {
```
Comment (manstis) [inline src/app/Instance/CreateInstance/machines/createInstanceMachine.ts:267]: Any reason not to rename this `_context` too (for consistency)?

File: src/app/Instance/CreateInstance/machines/createInstanceMachine.ts [hunk 7]
```diff
@@ -269,32 +282,42 @@ const createInstanceMachine = createMachine(
...
       fieldInvalid: send("fieldInvalid"),
       triggerSubmit: send("submit"),
     },
     guards: {
-      isNameEmpty: ({ name }) => name === undefined || name.trim().length === 0,
       isNameValid: ({ name, creationError }) =>
         name !== undefined &&
         name.trim().length > 0 &&
         creationError !== "name-taken",
+      isErrorHandlerValid: (context) => {
```
Comment (manstis) [inline src/app/Instance/CreateInstance/machines/createInstanceMachine.ts:318]: Any reason not to rename this `_context` too (for consistency)?

**Review by manstis — APPROVED** [review]
LGTM :+1: 

(I've also reinstated my _minor_ comments.. feel free to ignore them).

File: src/app/Instance/CreateInstance/machines/createInstanceMachine.ts (hunk 4 above)
Comment (kelvah) [reply src/app/Instance/CreateInstance/machines/createInstanceMachine.ts:283]: I only added the trailing underscore where `context` is not used to silence the `no-unused-vars` lint rule. (I've tried to remove it entirely too but it seems `assign` can't live without it).

Here the context is used a couple of lines below so the underscore is not necessary.

This same answer is valid for other places where I didn't rename it.

**Review by vpellegrino — APPROVED** [review]
Thanks @kelvah

**Comment by vpellegrino** [comment]
@Srihari1192 would you please give priority to this PR? My current task strictly depends on this. Thank you!

**Review by Srihari1192 — APPROVED** [review]

**Pull request merged**